portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java

```diff
@@ -0,0 +1,171 @@
+          lastExaminedTimeRange = currentRange;
+        } else {
+          // If lastExaminedTimeRange is later than the current range,
+          // add the current range to the new list of all ranges.
+          // The lastExaminedTimeRange pointer shouldn't move.
+          newTimeRanges.add(currentRange);
+        }
+      }
+
+      // If current time range is the last element in the allTimeRanges list,
+      // add the last examined time range to the new list.
+      if (i == allTimeRanges.size() - 1) {
+        newTimeRanges.add(lastExaminedTimeRange);
+      }
+    }
+
+    // Finally, set the global variable allTimeRanges to this newly built list of time ranges.
+    allTimeRanges = newTimeRanges;
+    Collections.sort(allTimeRanges, TimeRange.SORT_BY_TIME_RANGE_START_TIME_ASCENDING);
+  }
+
+  /** Helper method for merging two time ranges. This method is public so that it can be tested. */
+  public static TimeRange mergeTwoTimeRanges(TimeRange a, TimeRange b) {
+    Instant newTimeRangeStart;
+    Instant newTimeRangeEnd;
+
+    // The new time range after merging should have the earlier start time
+    // and the later end time among the two overlapping time ranges.
+    if (a.start().isBefore(b.start())) {
+      newTimeRangeStart = a.start();
+    } else {
+      newTimeRangeStart = b.start();
+    }
+
+    if (a.end().isBefore(b.end())) {
+      newTimeRangeEnd = b.end();
+    } else {
+      newTimeRangeEnd = a.end();
+    }
+
+    return TimeRange.fromStartEnd(newTimeRangeStart, newTimeRangeEnd);
+  }
+
+  /** Returns an iterator for the list of all time ranges. */
+  public Iterator<TimeRange> getAllTimeRanges() {
+    return allTimeRanges.iterator();
+  }
+
+  /**
+   * Checks if a time range exists in the collection. For example, if [3:00 - 4:00] is in the
+   * collection, [3:00 - 3:30] is considered to exist as a time range in the collection.
+   */
+  public boolean hasTimeRange(TimeRange timeRangeToCheck) {
+    for (TimeRange timeRange : allTimeRanges) {
+      if (timeRange.contains(timeRangeToCheck)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Delete a time range from the list. Because the list of all time ranges are always kept to be
+   * pairwise disjoint, the potentially two new time ranges resulted from a deletion will not
+   * overlap with any other existing time ranges.
+   *
+   * <p>For example, if the list contains [3:00 - 4:00] and [5:00 - 6:00], deleting [3:15 - 3:30]
+   * will result in [3 - 3:15] and [3:30 - 4] as new time ranges.
+   *
+   * <p>Another example for deleting overlapping time ranges: if [3 - 4] and [5 - 6] are in the
+   * original list, deleting [3:30 - 5:30] will result in two new ranges: [3 - 3:30] and [5:30 - 6].
+   */
+  public void deleteTimeRange(TimeRange timeRangeToDelete) {
+    List<TimeRange> newTimeRanges = new ArrayList<TimeRange>();
+
+    for (TimeRange currentRange : allTimeRanges) {
```

Comment:
What about the case where currentRange is completely covered by timeRangeToDelete?

Reply:
I think in that case, none of the two if statements inside the `if overlaps` gets executed, so the current time range that is covered by the range to delete is essentially ignored at this iteration of the for loop. I added a test case for this scenario.

Reply:
I see. Okay thanks.